docs/book/src/developer/testing.md

```diff
@@ -1,67 +1,157 @@
-For instance, after running e2e tests once with `SKIP_RESOURCE_CLEANUP=true`, the created management cluster can be used for the following e2e test runs by setting up `KUBECONFIG` variable.
+See [e2e development] for more information on developing e2e tests for CAPI and external providers.
 
-### Running tests
+## Running the end-to-end tests
 
 `make docker-build-e2e` will build the images for all providers that will be needed for the e2e test.
 
 `make test-e2e` will run e2e tests by using whatever provider images already exist on disk.
 After running `make docker-build-e2e` at least once, this can be used for a faster test run if there are no provider code changes.
 
-**Examples**
+Additionally, `test-e2e` target supports the following env variables:
 
-* `make docker-build-e2e`
+- `GINKGO_FOCUS` to set ginkgo focus (default empty - all tests)
+- `GINKGO_NODES` to set the number of ginkgo parallel nodes (default to 1)
+- `E2E_CONF_FILE` to set the e2e test config file (default to ${REPO_ROOT}/test/e2e/config/docker-dev.yaml)
+- `ARTIFACTS` to set the folder where test artifact will be stored (default to ${REPO_ROOT}/_artifacts)
+- `SKIP_RESOURCE_CLEANUP` to skip resource cleanup at the end of the test (useful for problem investigation) (default to false)
+- `USE_EXISTING_CLUSTER` to use an existing management cluster instead of creating a new one for each test run (default to false)
+- `GINKGO_NOCOLOR` to turn off the ginko colored output (default to false)
 
-* `make test-e2e GINKGO_FOCUS='upgrade' SKIP_RESOURCE_CLEANUP=true USE_EXISTING_CLUSTER=false`
+## Quick reference
 
-See [e2e development] for more information on developing e2e tests for CAPI and external providers.
+### `envtest`
+
+[envtest] is a testing environment that is provided by the [controller-runtime] project. This environment spins up a
+local instance of etcd and the kube-apiserver. This allows test to be executed in an environment very similar to a
+real environment. 
+
+Additionally, in Cluster API there is a set of utilities under [test/helpers] that helps developers in setting up
+a [envtest] ready for Cluster API testing, and most specifically:
+
+- With the required CRDs already pre-configured.
+- With all the Cluster API webhook pre-configured, so there are enforced guarantees about the semantic accuracy
+  of the test objects you are going to create.
+  
+Most notably, [envtest] does not provides only a real API server to user during test, but it offers the opportunity
+to configure one or more controllers to run against the test cluster; by using this feature it is possible to use
+[envtest] for developing Cluster API integration tests.
+
+Please note that, because [envtest] uses a real kube-apiserver that is shared across many tests, the developer
+should take care of ensuring each test run in isolation from the others, by:
+ 
+- Creating objects in separated namespaces.
+- Avoiding object name conflict.
+- Cleaning up objects created by each test.
```

Comment:
I'm guessing initially we will have an area of the codebase with tests written in the preferred style. This will probably result in creation of additional helpers that will help manage this overhead. These helpers should then be used everywhere else in order to provide a consistent test codebase.

Reply:
This makes totally sense